### PR TITLE
React 15 - Updates most dependencies to the latest version

### LINKED ISSUE #2: Update to use the latest dependencies
Use the latest version of React, React Router and more.

### PR DESCRIPTION
**Pull request by dlmr**
Also the examples are updated to reflect this change.

I have not updated React Helmet to the latest version since it was a bit to many questions about what has changed in the latest version.

I think we can consider that this closes #2.

### CODE REVIEW

File: packages/roc-package-web-app-react/package.json [hunk 1]
```diff
@@ -31,27 +31,22 @@
...
+    "history": "^2.1.0",
+    "nunjucks": "~2.4.2",
+    "pretty-error": "~2.0.0",
     "react-helmet": "~2.3.1",
-    "react-proxy": "1.1.8",
-    "react-redux": "~4.0.0",
-    "react-router": "~1.0.0",
+    "react-redux": "~4.4.5",
+    "react-router": "~2.3.0",
+    "react-router-redux": "4.0.2",
```
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/package.json]: Range or intended fixed version?

File: packages/roc-package-web-app-react/package.json [hunk 2]
```diff
@@ -31,27 +31,22 @@
...
     "react-server-status": "~1.0.0",
-    "redux": "~3.0.2",
-    "redux-devtools": "v3.0.0-beta-3",
-    "redux-devtools-dock-monitor": "1.0.0-beta-3",
-    "redux-devtools-log-monitor": "1.0.0-beta-3",
-    "redux-logger": "~2.1.3",
-    "redux-simple-router": "~0.0.8",
-    "redux-thunk": "~1.0.0",
+    "redial": "0.4.1",
```
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/package.json]: Range?

File: packages/roc-package-web-app-react/package.json [hunk 3]
```diff
@@ -31,27 +31,22 @@
   "dependencies": {
     "debug": "~2.2.0",
     "error": "~7.0.2",
-    "history": "~1.17.0",
-    "nunjucks": "~2.1.0",
-    "pretty-error": "~1.2.0",
-    "react-a11y": "~0.2.6",
-    "react-fetcher": "~0.2.0",
+    "history": "^2.1.0",
```
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/package.json:34]: `react-router@v2` bundles `history`, do we need an explicit dependency?

File: packages/roc-package-web-app-react/app/server/router.js [hunk 4]
```diff
@@ -33,7 +34,10 @@ export default function routes({ createRoutes, createStore, stats, dist }) {
                 this.status = 200;
                 this.body = renderPage();
             } else {
-                const store = createStore ? createStore() : null;
+                const { store, history, url } = setupForRender(createStore, this.url);
+
+                // Give Koa middlewares a chance to interact with the reduxStore
+                // This can be used to dynamically pass some data to the client.
```
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: I'd like to move this below `yield next`, to ensure we only create the store if we actually need it

File: packages/roc-package-web-app-react/package.json (hunk 1 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/package.json]: That should be a `~`, I have changed my default setting for npm and forgot to update this.

File: packages/roc-package-web-app-react/package.json (hunk 2 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/package.json]: That should be a `~`, I have changed my default setting for npm and forgot to update this.

File: packages/roc-package-web-app-react/app/server/render.js [hunk 5]
```diff
@@ -99,13 +96,11 @@ export function reactRender(url, createRoutes, store, renderPage, staticRender =
                     };
                 }
 
-                getPrefetchedData(components, locals)
+                trigger('fetch', components, locals)
                     .then(() => {
-                        let component = <RoutingContext {...renderProps} />;
+                        let component = <RouterContext {...renderProps} />;
```
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/render.js:101]: Any chance of, instead of triggering `'fetch'`, taking in a customizable function that can trigger whatever it needs. The default implementation, if not overridden, runs this triggering logic. If we go that route, we allow end users to hook into the server side logic and define additional fetching methods, such as optional data fetches that are allowed to fail and also triggering the deferred data fetches on the server. In this way, we can defer data fetching on the client, allowing the router to render the new route, while still allowing the server to fetch "all" data regardless of whether it happens before or after router transition on the client

File: packages/roc-package-web-app-react/package.json (hunk 3 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/package.json:34]: Yes, we need it because we are using a custom `history`. 

https://github.com/rocjs/roc-package-web-app-react/blob/feature/react-15/packages/roc-package-web-app-react/app/client/create-client.js#L8

File: packages/roc-package-web-app-react/app/server/router.js (hunk 4 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/router.js:40]: If we do this Koa middlewares will not have access to the `store`.

I suggest that we instead make it possible to define middlewares that run before the React renderer and after. The ones that run after then can use the `store` and the ones before can set `this.body` that we can use for an additional check in the React renderer.

Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: Good point!

Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: Or could we run the `match` before creating the store? That way we know if we're in the loop at all or not

Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/router.js:40]: I do not think so unfortunately, this because of two reasons. I hope that I'm missing something here since this would be a much better "API".

Firstly we need to do `syncHistoryWithStore` before we use the `history` instance in `match`. Maybe this can be avoided if we instead manually update the store to match the whatever state we get with this method.

Secondly we are currently passing in the `store` to the function that creates the routes. I think we want this to be able to use things like `onEnter` to check if the user is authenticated and if not perform a redirect.

File: packages/roc-package-web-app-react/app/server/render.js (hunk 5 above)
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/render.js:101]: If we do this though, that should be a separate PR

Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/render.js:101]: I'm not sure what you mean here, what do you want to achieve? _I have created a suggested API below that might offer what is requested._

> additional fetching methods, such as optional data fetches that are allowed to fail and also triggering the deferred data fetches on the server.

Like creating additional things that needs to be completed before rendering the page? So an API for this could look like this?

``` js
const extra = userDefinedThings.map((userDefinedThing) => userDefinedThing({
  dispatch,
  getState,
  location,
  params,
  components
})

Promise.all([trigger, ...extra]).then(() => { //...
```

> triggering the deferred data fetches on the server

Why would we want to perform the `defer` data fetching on the server?

---

I have also looked into some other ways to modify what the server does based on our discussion from earlier. This makes it possible to 
- Define extra things on `locals` that is used for `trigger()`
  _Not sure that this is a good idea since we might have a miss match from what the client have and more checks would need to be performed the function in both contexts_
- Use things like `react-side-effects` and other things that should modify the template.

``` js
// "Pseudo" code
match(..., () => {
  // Do checks about if the path was "valid"
  const { locals } = runBeforeRender({ store }) // Can perform dispatch and give back locals

  const page = ReactDOM.renderToString()

  const { templateProps } = runAfterRender({ getState }) // Would be possible to modify store here but is that a good idea?

  return renderPage(page, templateProps)
}) 
```

In this example runAfterRender will be able to do for instance `.rewind()` on some singleton thing.

Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/render.js:101]: Yeah sorry, I was writing on my phone, so I might have been quite confusing. Ok, so in Hyperion, we have `fetchData` and `fetchDataDeferred`, right? Now actually **both** of them run on the server as well as the client. When we get a `ROUTER_DID_CHANGE` action, we run the `fetchData`-method. When this action resolves, then we dispatch the original `ROUTER_DID_CHANGE` action and the page renders for the new location. Then `fetchDataDeferred` actions are run, and they in turn can update the state by dispatching actions, causing the page to re-render as needed.

_On the server_:
- We listen to the resolution of the data fetches (a `Promise` is made that resolved once both `fetchData` and `fetchDataDeferred` is complete)
- When the data fetching is complete, we render the page and all data is available

_On the client_:
- No special handling. `@connect` ensures that the page is up to date.

### Pros and cons of this approach

**Pros**
- The article page uses `fetchDataDeferred` to allow us to immediately render the Article page when clicking on a teaser from a section page. This gives the user immediate feedback that something is happening. It also gives us a chance to render a Facebook-style Article placeholder page while the article data is fetched
- We fetch all data needed for rendering on the server
- Allowing the client side to fetch data before and/after router transition is pretty sweet. Perhaps a better name for the methods would be `fetchDataBeforeRouterTransition` and `fetchDataAfterRouterTransition`

**Cons**
- If `fetchDataDeferred` is used, then the route handler must handle the case where the data is not yet available
- If we _actually do want to skip loading_ some data on the server, then we have no good way of triggering client side only data fetches.
- We have no concept of "optional" data that is allowed to fail on the server and retries on the client

Now, I actually like what you're doing in Roc better, but by allowing some way of running additional functions we can make it as flexible as the end user needs it to be.

File: packages/roc-package-web-app-react/app/server/router.js (hunk 4 above)
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: > Firstly we need to do syncHistoryWithStore before we use the history instance in match. Maybe this can be avoided if we instead manually update the store to match the whatever state we get with this method.

Hmm. You're right. Sounds a bit hacky. Is there no way to get `redux-simple-router` to dispatch the current location to the store if `history` already has a location when syncing?

> Secondly we are currently passing in the store to the function that creates the routes. I think we want this to be able to use things like onEnter to check if the user is authenticated and if not perform a redirect.

Right, so basically we want to let the eventual `onEnter`-hooks to get a reference to `getState()` and maybe even `dispatch()`?

File: packages/roc-package-web-app-react/app/server/render.js (hunk 5 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/render.js:101]: I see, now it makes sense! 🙂

Most of this is already available through the current structure, just a bit different in the APIs and how Redux is used. Instead of managing transition with Redux on the client side we are using a `RouterContext`. This is since version 1.0 of React Router the recommended way to do this sort of things. Currently it is coded to load `fetch` and `defer` simultaneously but make the transition as soon as the `fetch` is completed **if** `blocking` in settings is defined to `true` (default). _I see that this might not be what one wants, it might be more logical to first completely load `fetch` and then `defer` if `blocking` is `true`.

The server will, currently, only prefetch `fetch` as you have already mentioned. The reasoning for this is that we might want to render the page as soon as possible and then being able to _defer_ some data fetching to the client can be beneficial. I however agree with you that in some cases this is not the expected/wanted behaviour and I suggest the following.

A setting that defines what Redial _"names"_ that should be fetched on the server before rendering. 

``` js
// Not the actual syntax
settings.runtime.fetchOnServer = ['fetch', 'defer', 'only-server']
```

Then we probably want to do something similar on the client but for before and after route transition, if `blocking` is `true`.

``` js
// Not the actual syntax
settings.runtime.fetchOnClient = {
  blocking: ['fetch'],
  nonBlocking: ['defer', 'client-only']
}
```

---

Since we actually can create anything it might be interesting to be able to define something similar to lifecycle things. An example of this would be `done` that maybe should run after all the other "blocking" things have been completed. This would allow these functions to have an updated state from the previous ones.

File: packages/roc-package-web-app-react/app/server/router.js (hunk 4 above)
Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/router.js:40]: > Is there no way to get redux-simple-router to dispatch the current location to the store if history already has a location when syncing?    

We might be able to use [these](https://github.com/reactjs/react-router-redux/#pushlocation-replacelocation-gonumber-goback-goforward) to do it but I'm not sure how that works with `basename` that we are now supporting.

> Right, so basically we want to let the eventual onEnter-hooks to get a reference to getState() and maybe even dispatch()?  

I believe that would be nice to have since it would be trivial to do [this](https://github.com/reactjs/react-router/blob/master/examples/auth-flow/app.js#L133).

File: packages/roc-package-web-app-react/app/server/render.js (hunk 5 above)
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/render.js:101]: This will be awesome! Let's leave this discussion for now in this PR, because it is should have its own scope, and has nothing to do with updating dependencies

I'll create an issue with your proposed solution, then we can take it from there, ok?

Comment (dlmr) [inline packages/roc-package-web-app-react/app/server/render.js:101]: Yeah let's do that!

Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/render.js:101]: See #8

File: packages/roc-package-web-app-react/app/server/router.js (hunk 4 above)
Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: Yeah, very nice indeed.

Anyway, this is also a separate issue and not related to this PR.

Comment (PAkerstrand) [inline packages/roc-package-web-app-react/app/server/router.js:40]: Issue created here: #9

**Comment by lizell** [comment]
Awesome discussions, guys!

File: packages/roc-package-web-app-react/app/server/render.js [hunk 6]
```diff
@@ -79,7 +76,7 @@ export function reactRender(url, createRoutes, store, renderPage, staticRender =
                 } else if (!renderProps) {
                     log('No renderProps, most likely the path does not exist');
                     return resolve({
-                        status: 500,
+                        status: 404,
```
Comment (lizell) [inline packages/roc-package-web-app-react/app/server/render.js:79]: :star:

**Comment by lizell** [comment]
:+1:

**Comment by PAkerstrand** [comment]
👍

**Comment by andreasrs** [comment]
Great that we got some issues created for moving onward on those things!

In terms of that this PR; I think it's pretty ready to be merged.

Anyone have some outstanding things they want to address?
Edit: Just saw :+1: from @lizell and @PAkerstrand. 
@DLMR ?

Not 100% sure of best way to move toward a release here. It could be that it's worthy a bump to _ beta_  pre-release because of the new major dependencies that will require changes within projects.

Will carry on the SemVer discussion outside of this PR.

**Comment by dlmr** [comment]
> Not 100% sure of best way to move toward a release here. It could be that it's worthy a bump to _ beta_ pre-release because of the new major dependencies that will require changes within projects.

I think we should stay on `alpha`, bumping it to `beta` means nothing in this context.

I want to update React Helmet and will be able to do that later today, so maybe keep this PR active until I get around to do that.

**Comment by andreasrs** [comment]
Alright. I guess a changelog on the pre-release will have to suffice.
Let's have a chat about getting the extensions out of pre-release soon.
